### ISSUE TITLE
Can't enable FIPS on ubuntu pro machine

### ISSUE THREAD

**Issue by chenhongqiao** [canonical/ubuntu-advantage-client#1087]
When trying to enable FIPS on a AWS Ubuntu Pro instance by prompting "ua enable fips", it failed and returned "GPG key '/usr/share/keyrings/ubuntu-fips-keyring.gpg' not found"

**Comment by mitechie** [comment]
Do you have the full output please for this, please?

Thanks!

**Comment by cnewcome** [comment]
[ua-debug-enable-fips.txt](https://github.com/canonical/ubuntu-advantage-client/files/4934051/ua-debug-enable-fips.txt)

I ran into this issue as well. Here is the full debug output of my "ua --debug enable fips" command. I get the same keyring error.

Please let me know if you need more info. Feel free to ping me directly on MM.

**Comment by cnewcome** [comment]
```
root@ip-....:/etc/apt/auth.conf.d# ua status
SERVICE       ENTITLED  STATUS    DESCRIPTION
cc-eal        yes       n/a       Common Criteria EAL2 Provisioning Packages
cis-audit     no        —         Center for Internet Security Audit Tools
esm-apps      yes       enabled   UA Apps: Extended Security Maintenance
esm-infra     yes       enabled   UA Infra: Extended Security Maintenance
fips          yes       disabled  NIST-certified FIPS modules
fips-updates  yes       disabled  Uncertified security updates to FIPS modules
livepatch     yes       enabled   Canonical Livepatch service

Enable services with: ua enable <service>

                Account: ....
           Subscription: ....
            Valid until: n/a
Technical support level: essential


root@ip-172-31-85-184:/etc/apt/auth.conf.d# ua version
24.2~18.04.1
```

**Comment by roscoecairney** [comment]
Seeing the same on a freshly launched Ubuntu Pro in AWS

```
root@ip-10-142-18-65:/var/snap/amazon-ssm-agent/2012# dpkg -l | grep advantage
ii  ubuntu-advantage-pro             24.2~18.04.1                        amd64        utilities and services for Ubuntu Pro images
ii  ubuntu-advantage-tools           24.2~18.04.1                        amd64        management tools for Ubuntu Advantage
```
```
root@ip-10-142-18-65:/var/snap/amazon-ssm-agent/2012# ua status
SERVICE       ENTITLED  STATUS    DESCRIPTION
cc-eal        yes       n/a       Common Criteria EAL2 Provisioning Packages
cis-audit     no        —         Center for Internet Security Audit Tools
esm-apps      yes       enabled   UA Apps: Extended Security Maintenance
esm-infra     yes       enabled   UA Infra: Extended Security Maintenance
fips          yes       disabled  NIST-certified FIPS modules
fips-updates  yes       disabled  Uncertified security updates to FIPS modules
livepatch     yes       enabled   Canonical Livepatch service

Enable services with: ua enable <service>

                Account: xxxxxxxxxxxx
           Subscription: xxxxxxxxxxxx
            Valid until: n/a
Technical support level: essential
```
```
root@ip-10-142-18-65:/var/snap/amazon-ssm-agent/2012# ua enable fips
One moment, checking your subscription first
GPG key '/usr/share/keyrings/ubuntu-fips-keyring.gpg' not found
```

**Comment by drimer** [comment]
I've been having the same problem.

```
root@ip-10-142-16-239:/# ua status
SERVICE       ENTITLED  STATUS    DESCRIPTION
cc-eal        yes       n/a       Common Criteria EAL2 Provisioning Packages
cis-audit     no        —         Center for Internet Security Audit Tools
esm-apps      yes       enabled   UA Apps: Extended Security Maintenance
esm-infra     yes       enabled   UA Infra: Extended Security Maintenance
fips          yes       disabled  NIST-certified FIPS modules
fips-updates  yes       disabled  Uncertified security updates to FIPS modules
livepatch     yes       enabled   Canonical Livepatch service

Enable services with: ua enable <service>

                Account: ##
           Subscription: ##
            Valid until: n/a
Technical support level: essential
```

And when I try to enable it, I also get the same:

```
root@ip-10-142-16-239:/# ua enable fips
One moment, checking your subscription first
GPG key '/usr/share/keyrings/ubuntu-fips-keyring.gpg' not found
```

Is there any  way to work around this while it gets fixed? I came across this script: https://github.com/canonical/ubuntu-advantage-client/pull/579/files#diff-294d733f999f5f4f12c570f6596751a4 and tried running it to get new GPG keys, but that hasn't worked either.

**Comment by jbruni** [comment]
I've been able to get the missing key with this command:

```sh
gpg --keyring "/usr/share/keyrings/ubuntu-fips-keyring.gpg" --no-default-keyring --keyserver keyserver.ubuntu.com --recv-keys C1997C40EDE22758
```

**Comment by roscoecairney** [comment]
Thanks jbruni. I've run your command, but receive this result.
```
root@ip-10-142-18-55:/var/snap/amazon-ssm-agent/1566# gpg --keyring "/usr/share/keyrings/ubuntu-fips-keyring.gpg" --no-default-keyring --keyserver keyserver.ubuntu.com --recv-keys C1997C40EDE22758
gpg: keybox '/usr/share/keyrings/ubuntu-fips-keyring.gpg' created
gpg: failed to create temporary file '/home/ssm-user/.gnupg/.#lk0x000055a70845ec70.ip-10-142-18-55.3624': No such file or directory
gpg: connecting dirmngr at '/home/ssm-user/.gnupg/S.dirmngr' failed: No such file or directory
gpg: keyserver receive failed: No dirmngr
```

However, I think `ua enable fips` should not require me to perform any manual steps. If it requires keys, then it should add them.

**Comment by mitechie** [comment]
This should be shipped with the updated ua client that is supported on xenial/bionic. 

https://github.com/canonical/ubuntu-advantage-client/tree/master/keyrings

**Comment by rcj4747** [comment]
On Xenial the workaround from @jbruni (https://github.com/canonical/ubuntu-advantage-client/issues/1087#issuecomment-670984444) was sufficient, but on Bionic `ua attach fips`  fails even after creating /usr/share/keyrings/ubuntu-fips-keyring.gpg.  Here's debug output:

```
$ sudo gpg --keyring "/usr/share/keyrings/ubuntu-fips-keyring.gpg" --no-default-keyring --keyserver keyserver.ubuntu.com --recv-keys C1997C40EDE22758
...
$ sudo ua --debug enable fips
DEBUG: Executed with sys.argv: ['/usr/bin/ua', '--debug', 'enable', 'fips']
DEBUG: Reading file: /var/lib/ubuntu-advantage/private/machine-token.json
One moment, checking your subscription first
DEBUG: Reading file: /etc/os-release
DEBUG: Reading file: /etc/os-release
DEBUG: Reading file: /etc/os-release
DEBUG: Reading file: /etc/os-release
DEBUG: Reading file: /etc/os-release
DEBUG: Reading file: /etc/os-release
DEBUG: Reading file: /etc/machine-id
DEBUG: Reading file: /etc/os-release
DEBUG: URL [POST]: https://contracts.canonical.com/v1/contracts/...
DEBUG: URL [POST] response: https://contracts.canonical.com/v1/contracts/...
Date: Thu, 20 Aug 2020 20:50:01 GMT
Content-Type: application/json
Transfer-Encoding: chunked
Connection: close
Vary: Accept-Encoding
Strict-Transport-Security: max-age=15724800; includeSubDomains
...
DEBUG: Writing file: /var/lib/ubuntu-advantage/private/machine-token.json
DEBUG: Reading file: /var/lib/ubuntu-advantage/private/machine-token.json
DEBUG: Reading file: /etc/os-release
DEBUG: Reading file: /etc/os-release
DEBUG: Reading file: /etc/os-release
DEBUG: Reading file: /etc/os-release
DEBUG: Reading file: /etc/os-release
DEBUG: Reading file: /etc/os-release
DEBUG: Reading file: /etc/os-release
DEBUG: Reading file: /etc/os-release
DEBUG: Reading file: /etc/os-release
DEBUG: Reading file: /etc/os-release
DEBUG: Reading file: /etc/os-release
DEBUG: Reading file: /etc/os-release
DEBUG: Ran cmd: apt-cache policy, rc: 0 stderr: b''
DEBUG: Reading file: /etc/os-release
DEBUG: Reading file: /etc/os-release
DEBUG: No specific resourceToken present. Using machine token as FIPS credentials
DEBUG: Reading file: /etc/os-release
DEBUG: Writing file: /etc/apt/preferences.d/ubuntu-fips-bionic
DEBUG: Reading file: /etc/os-release
DEBUG: Ran cmd: apt-cache policy, rc: 0 stderr: b''
DEBUG: Writing file: /etc/apt/sources.list.d/ubuntu-fips-bionic.list
DEBUG: Reading file: /etc/apt/auth.conf.d/90ubuntu-advantage
DEBUG: Writing file: /etc/apt/auth.conf.d/90ubuntu-advantage
DEBUG: Exporting GPG key /usr/share/keyrings/ubuntu-fips-keyring.gpg
Updating package lists
DEBUG: Failed running command 'apt-get update' [exit(100)]. Message: W: http://us-east-1.ec2.archive.ubuntu.com/ubuntu/dists/bionic/InRelease: The key(s) in the keyring /etc/apt/trusted.gpg.d/ubuntu-fips-keyring.gpg are ignored as the file has an unsupported filetype.
W: http://us-east-1.ec2.archive.ubuntu.com/ubuntu/dists/bionic-updates/InRelease: The key(s) in the keyring /etc/apt/trusted.gpg.d/ubuntu-fips-keyring.gpg are ignored as the file has an unsupported filetype.
W: http://us-east-1.ec2.archive.ubuntu.com/ubuntu/dists/bionic-backports/InRelease: The key(s) in the keyring /etc/apt/trusted.gpg.d/ubuntu-fips-keyring.gpg are ignored as the file has an unsupported filetype.
W: https://esm.ubuntu.com/apps/ubuntu/dists/bionic-apps-security/InRelease: The key(s) in the keyring /etc/apt/trusted.gpg.d/ubuntu-fips-keyring.gpg are ignored as the file has an unsupported filetype.
W: https://esm.ubuntu.com/apps/ubuntu/dists/bionic-apps-updates/InRelease: The key(s) in the keyring /etc/apt/trusted.gpg.d/ubuntu-fips-keyring.gpg are ignored as the file has an unsupported filetype.
W: https://esm.ubuntu.com/infra/ubuntu/dists/bionic-infra-security/InRelease: The key(s) in the keyring /etc/apt/trusted.gpg.d/ubuntu-fips-keyring.gpg are ignored as the file has an unsupported filetype.
W: https://esm.ubuntu.com/infra/ubuntu/dists/bionic-infra-updates/InRelease: The key(s) in the keyring /etc/apt/trusted.gpg.d/ubuntu-fips-keyring.gpg are ignored as the file has an unsupported filetype.
W: https://esm.ubuntu.com/fips/ubuntu/dists/bionic/InRelease: The key(s) in the keyring /etc/apt/trusted.gpg.d/ubuntu-fips-keyring.gpg are ignored as the file has an unsupported filetype.
W: GPG error: https://esm.ubuntu.com/fips/ubuntu bionic InRelease: The following signatures couldn't be verified because the public key is not available: NO_PUBKEY C1997C40EDE22758
E: The repository 'https://esm.ubuntu.com/fips/ubuntu bionic InRelease' is not signed.
W: http://ppa.launchpad.net/canonical-server/ua-client-premium/ubuntu/dists/bionic/InRelease: The key(s) in the keyring /etc/apt/trusted.gpg.d/ubuntu-fips-keyring.gpg are ignored as the file has an unsupported filetype.
W: http://security.ubuntu.com/ubuntu/dists/bionic-security/InRelease: The key(s) in the keyring /etc/apt/trusted.gpg.d/ubuntu-fips-keyring.gpg are ignored as the file has an unsupported filetype.
```
